Arbitrary self types v2: shadowing semver break. (#15117)
rust-lang/cargo: src/doc/src/reference/semver.md

safe is also a minor change, while the same is not true for associated
function on traits (see [any change to trait item signatures](#trait-item-signature)).

### Major: add a potentially shadowing method {#fn-add-potentially-shadowing-method}

If you have a type which implements `Deref<Target=T>`, you must not add methods
which may "shadow" methods in `T`. This can lead to unexpected changes in

> program behavior.

Reviewer: "must not" is too strong.

This is intentionally done and relied on in a number of cases.

Author: Would you agree with the statement if I say:

If you have a type which implements  `Deref<Target=T>` **for an arbitrary user-controlled `T`**, you must not add methods which may "shadow" methods in `T`.

I'm quite keen to give a precise hard rule because it will make it easier for the `cargo semver` folks, if nothing else.

If you still think that's not a true statement, could you give examples of where this is intentionally done and is OK? Some of these cases will turn into hard errors when/if https://github.com/rust-lang/rust/pull/135881 lands.

Author: @obi1kenobi @workingjubilee I'd appreciate a little more advice here about how to resolve this concern.

Reviewer 2: As a possible path forward, in an effort to save everyone else's valuable time:
```suggestion
If you have a type which implements `Deref<Target=T>`, adding methods to that type
which may "shadow" methods in `T` can cause breaking changes in program behavior.
```

Reviewer: So, on stable, you can already have the case of function "overloading" done via using the fact that inherent implementations override trait implementations, and an existing pattern in-use in the ecosystem is to have `Trait` and then have `SometimesShadows<T>` which has
```rust
impl<T> Deref for SometimesShadows<T> 
where
    T: Trait,
{
    type Target = T;

    fn deref(&self) -> &Self::Target {
        todo!()
    }
}
```
and then for `SometimesShadows<T>` you deliberately "shadow" certain otherwise-deref'd-to functions, so that you have an API where you for various reasons deliberately arrange to sometimes call functions on either `Trait`, `T`, or `SometimesShadows<T>`. Then usually you repeat this for multiple different possible `SometimesShadows<T>`s, where each instance of this wrapper represents something and the deliberate override happens in certain cases. Then you have some code that gets macro-expanded in a monomorphic fashion and is the "real" driver for all of this.

This is a complicated abstraction, as you can tell, but I have seen it done, in particular to enable a sort of cross-language reflection scheme. And yes, it intentionally worked with what were potentially arbitrary user-provided `T`s.

Reviewer: Basically the underlying truth of the matter is:

People do what is sometimes the most bizarre stuff *quite on purpose*, and while some of these decisions indeed are "breaking changes", they're quite deliberate when they are designed this way. Indeed, simply skipping over "must not" or "should not" is probably enough. Just tell people "this is a major change, this is not, and if you want to make breaking changes? uh, enjoy that, I guess? no refunds?"

If you have a type which implements `Deref<Target=T>`, you must not add methods
which may "shadow" methods in `T`. This can lead to unexpected changes in
program behavior.

```rust,ignore,run-fail
// MAJOR CHANGE

///////////////////////////////////////////////////////////
// Before
#[derive(Clone, Copy)]
pub struct MySmartPtr<T>(pub T);

impl<T> core::ops::Deref for MySmartPtr<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

///////////////////////////////////////////////////////////
// After
#[derive(Clone, Copy)]
pub struct MySmartPtr<T>(pub T);

impl<T> core::ops::Deref for MySmartPtr<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> MySmartPtr<T> {
    pub fn method(self) -> usize {
        2
    }
}

///////////////////////////////////////////////////////////
// Example usage that will break.
use updated_crate::MySmartPtr;

struct SomeStruct;

impl SomeStruct {
    fn method(&self) -> usize {
        1
    }
}

fn main() {
    let mut ptr = MySmartPtr(SomeStruct);
    assert_eq!(ptr.method(), 1);
}
```

Note that the shadowing and shadowed methods receive `self`

> slightly differently: `&self` and `&mut self`.

Reviewer 3: The code above doesn't use `&mut self`.

Author: Oops yes - my edits to code and prose got out of sync - I'll fix it tomorrow.

Author: Fixed.

That's because Rust searches for methods first by value, then by `&`, then

> by `&mut T`, then by `*const T`. Rust stops the search

Reviewer 3: Is this true that `*const T` is on the list?  That's not how we [document it](https://doc.rust-lang.org/nightly/reference/expressions/method-call-expr.html#method-call-expressions).

Author: Mention of `*const T` removed.

Reviewer 2: Even if it isn't in the reference, it would be super useful to have a completely-accurate writeup somewhere, for the sake of cargo-semver-checks. We implement lints as best we understand the domain, and any divergences between what we implement and what rustc actually does are experienced as bugs by our users.

So personally, I'd advocate for either a special case callout in the reference ("if the `self` type is `*mut T` ..."), or at least a comment here containing the SemVer implications of such a case. Knowing about it now can save cargo-semver-checks maintainers hours of trying to triage bugs and figure out the exact rule later 🙏

Reviewer 3: > In retrospect I was wrong to mention this here, and I'll remove it. It does slightly bother me that the reference is incomplete here, but as far as I know it always has been, and I assume somebody made a principled decision to keep it clear and comprehensible instead of describing every last detail.

That may or may not be true.  Typically we do want to describe every last user-visible behavioral detail unless we've decided to treat the behavior as not-guaranteed for some reason.  This doesn't strike me as the latter case offhand.  If the Reference isn't complete here, we probably just want to update it.

@ehuss might have some history and further context here.

Reviewer 3: That is to say, @adetaylor, we'd welcome a PR to the Reference that would take a shot at describing this better.

Author: I had a crack at a major rewrite of that bit of the reference in https://github.com/rust-lang/reference/pull/1725.
